### PR TITLE
PARQUET-438: Update RLE encoding tools and add unit tests from Impala

### PR DESCRIPTION
**Pull request by wesm**
This also changes the compiler options to use only SSE3. We can explore benchmarking to optionally enable SSE4 on supported systems for performance gains in the future. 

@nongli would it be OK if I remove the unused delta encodings (they depend on code that is not in Impala anymore, or never way)? Are these encodings used extensively, or primarily PLAIN / DICTIONARY encoding?

### CODE REVIEW

**Comment by majetideepak** [comment]
We need the RLE decoder to handle BIT_PACKED encoding as well. Can this code decode BIT_PACKED values ?

**Comment by nongli** [comment]
@wesm Which encoding are you talking about? Some of these are the 2.0 encodings which should not removed. Let's just add the utilities we need from Impala in here.

**Comment by wesm** [comment]
@majetideepak no, and it does not need to. You can just use a `BitReader` in that case, see https://github.com/cloudera/Impala/blob/cdh5-trunk/be/src/exec/hdfs-parquet-scanner.cc#L1247

To handle `BIT_PACKED` level encoding, like Impala we should insert a `LevelDecoder` abstraction which dispatches either to `RLE` or `BIT_PACKED` encoding as appropriate.

**Comment by wesm** [comment]
@nongli all of the `delta*` headers

**Comment by majetideepak** [comment]
`LevelDecoder` abstraction works as well.
I will combine PARQUET-169 and PARQUET-462 similar to the Impala's `LevelDecoder`

**Comment by wesm** [comment]
@majetideepak can you base that change on PARQUET-435 / #26 ? thanks

**Comment by majetideepak** [comment]
Yes! I will. Makes more sense.

**Comment by nongli** [comment]
@wesm we should not remove those. Those are the 2.0 encoding implementations. Impala doesn't support 2.0 yet.

**Comment by wesm** [comment]
Got it, thank you. I commented out the includes -- we can go back and reconcile the implementations when someone is ready to write unit tests. Perhaps by the time Impala will have implemented the encodings.

**Comment by majetideepak** [comment]
+1 LGTM

**Comment by majetideepak** [comment]
I plan to rebase PARQUET-169 and PARQUET-462 on #31

**Comment by wesm** [comment]
I have an incoming patch that refactors `ReadNewPage` -- but I don't think the conflict will be too bad

**Comment by wesm** [comment]
(per PARQUET-485)

File: src/parquet/util/rle-encoding.h [hunk 1]
```diff
@@ -93,170 +94,171 @@ class RleDecoder {
...
+    repeat_count_ = 0;
+    literal_count_ = 0;
+  }
+
+  /// Gets the next value.  Returns false if there are no more.
   template<typename T>
   bool Get(T* val);
 
- private:
+ protected:
```
Comment (majetideepak) [inline src/parquet/util/rle-encoding.h]: Should we keep this private ?

Comment (wesm) [inline src/parquet/util/rle-encoding.h]: Since this was a copy paste job I would rather not make any unnecessary changes -- I'm also exploring the possibility of making parquet-cpp a dependency of Impala and this would cause code breakage.

**Comment by wesm** [comment]
This has been rebased on PARQUET-496 and the build should be passing again

File: src/parquet/util/bit-stream-utils.inline.h [hunk 2]
```diff
@@ -83,13 +85,9 @@ inline bool BitWriter::PutVlqInt(uint32_t v) {
   return result;
 }
 
-inline bool BitWriter::PutZigZagVlqInt(int32_t v) {
```
Comment (nongli) [inline src/parquet/util/bit-stream-utils.inline.h]: Can you add these (zigzag) back? I think this makes the new decoders work.

You can leave a TODO for me to add tests.

Comment (wesm) [inline src/parquet/util/bit-stream-utils.inline.h]: Sure.

**Comment by wesm** [comment]
@nongli the signatures for `GetVlqInt` and `PutVlqInt` are different in Impala from what they were before this patch. I am not confident restoring the ZigZag functions and twiddling integer types without tests; let me know what you want to do

**Comment by nongli** [comment]
Rather than just removing it, just comment it out and i'll take a look.

**Comment by wesm** [comment]
Roger. Thanks

**Comment by wesm** [comment]
All set.

File: src/parquet/util/rle-encoding.h (hunk 1 above)
Comment (julienledem) [inline src/parquet/util/rle-encoding.h]: +1 on impala using parquet-cpp!

File: src/parquet/util/bit-util-test.cc [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+  EXPECT_EQ(BitUtil::RoundDownNumi32(33), 1);
+
+  EXPECT_EQ(BitUtil::RoundUpNumi64(63), 1);
+  EXPECT_EQ(BitUtil::RoundUpNumi64(64), 1);
+  EXPECT_EQ(BitUtil::RoundUpNumi64(65), 2);
+  EXPECT_EQ(BitUtil::RoundDownNumi64(63), 0);
+  EXPECT_EQ(BitUtil::RoundDownNumi64(64), 1);
+  EXPECT_EQ(BitUtil::RoundDownNumi64(65), 1);
+}
+
```
Comment (julienledem) [inline src/parquet/util/bit-util-test.cc]: nice

**Comment by julienledem** [comment]
LGTM. any more comments?

**Comment by majetideepak** [comment]
+1 LGTM

On Mon, Feb 1, 2016 at 7:18 PM, Julien Le Dem <notifications@github.com>
wrote:

> LGTM. any more comments?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/parquet-cpp/pull/31#issuecomment-178270207>.
>



-- 
regards,
Deepak Majeti

**Comment by julienledem** [comment]
I'll merge then :)
